provisioner/node_pools.go

```diff
@@ -123,13 +223,150 @@ func (p *AWSNodePoolProvisioner) Provision(ctx context.Context, nodePools []*api
 	return nil
 }
 
-func supportsT2Unlimited(instanceTypes []string) bool {
-	for _, instanceType := range instanceTypes {
-		if !strings.HasPrefix(instanceType, "t2.") {
-			return false
+func (p *KarpenterNodePoolProvisioner) kubectlExecute(ctx context.Context, args []string, stdin string) (string, error) {
+	token, err := p.tokenSource.Token()
+	if err != nil {
+		return "", err
+	}
+
+	args = append([]string{
+		"kubectl",
+		fmt.Sprintf("--server=%s", p.cluster.APIServerURL),
+		fmt.Sprintf("--token=%s", token.AccessToken),
+	}, args...)
+	if stdin != "" {
+		args = append(args, "-f", "-")
+	}
+
+	newApplyCommand := func() *exec.Cmd {
+		cmd := exec.Command(args[0], args[1:]...)
+		// prevent kubectl to find the in-cluster config
+		cmd.Env = []string{}
+		return cmd
+	}
+	var output string
+	applyManifest := func() error {
+		cmd := newApplyCommand()
+		if stdin != "" {
+			cmd.Stdin = strings.NewReader(stdin)
 		}
+		output, err = p.execManager.Run(ctx, p.logger, cmd)
+		return err
 	}
-	return true
+	err = backoff.Retry(applyManifest, backoff.WithMaxRetries(backoff.NewExponentialBackOff(), maxApplyRetries))
+	if err != nil {
+		return "", err
+	}
+	return output, nil
+}
+
+func (p *KarpenterNodePoolProvisioner) provisionNodePool(ctx context.Context, nodePool *api.NodePool, values map[string]interface{}) error {
+	template, err := p.generateNodePoolTemplate(nodePool, values, provisionersFileName)
+	if err != nil {
+		return err
+	}
+	if template == "" {
+		return nil
+	}
+	_, err = p.kubectlExecute(ctx, []string{"apply"}, template)
+	if err != nil {
+		return errors.Wrapf(err, "kubectl apply failed for node pool %s", nodePool.Name)
+	}
+	return nil
+}
+
+func (p *KarpenterNodePoolProvisioner) isKarpenterInstalled(ctx context.Context) (bool, error) {
+	gvr, err := kubernetes.ResolveKind(p.kubeMapper, crd)
+	if err != nil {
+		return false, err
+	}
+
+	_, err = p.kubeClient.Resource(gvr).Get(ctx, karpenterProvisionerResource, metav1.GetOptions{})
+	if err != nil {
+		serr, ok := err.(*apiErrors.StatusError)
+		if ok && serr.Status().Code == http.StatusNotFound {
+			return false, nil
+		}
+		return false, err
+	}
+	return true, nil
+}
+
+func (p *KarpenterNodePoolProvisioner) Reconcile(ctx context.Context, updater updatestrategy.UpdateStrategy) error {
+	karpenterPools := p.cluster.KarpenterPools()
+	installed, err := p.isKarpenterInstalled(ctx)
+	if err != nil {
+		return err
+	}
+	if !installed {
+		// skip if karpenter is not installed
+		p.logger.Infof("skipping reconcilation of karpenter node pools as karpenter is not installed")
+		return nil
```

Comment:
I unresolved this because I still see this as questionable

Reply:
isKarpenterEnabled checks if the feature flag is set to true.
I thought it would be simpler to skip the operation if the flag is false.

I will move the check into `clusterpy.go` only call reconcile if karpenter feature flag is enabled.

Reply:
done